Correct how meta-field is defined for pre 7.8 hits

mayya-sharipova opened the pull request:
We keep a static list of meta-fields: META_FIELDS_BEFORE_7_8
as it was before.
This is done to ensure the backwards compatability with pre 7.8 nodes.

Relates to #57771
Closes #57831

mayya-sharipova committed:
Correct how meta-field is defined for pre 7.8 hits

We keep a static list of meta-fields: META_FIELDS_BEFORE_7_8
as it was before.
This is done to ensure the backwards compatability with pre 7.8 nodes.

Closes #57831

cbuescher commented on server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:
```diff
@@ -813,11 +818,7 @@ public void close() throws IOException {
-        }
-        // if a node had Size Plugin installed, _size field should also be considered a meta-field
-        return fieldName.equals("_size");
+        return META_FIELDS_BEFORE_7_8.contains(fieldName);
```
If I remember correctly we only use `isMetadataFieldStatic` in a few places now where we read from older nodes. Should we simply make the collection public and remove the static method? Or rename method to something like "isMetaFieldBefore7dot8" or something, just to improve readability in the locations where this it is actually used.

mayya-sharipova replied:
@cbuescher Thanks for the feedback! I have removed `isMetadataFieldStatic` method, and made the collection public.

cbuescher reviewed (approved):
LGTM, I left a minor nit, feel free to ignore

mayya-sharipova committed:
Address Chrishoph's feedback

mayya-sharipova committed:
Ensure that META_FIELDS_BEFORE_7DOT8 is used to create test hit instances

SearchHitsTests.testSerializationPre6_6_0 will fail
if you use different meta-fields that were before

cbuescher commented on server/src/test/java/org/elasticsearch/index/get/DocumentFieldTests.java:
```diff
@@ -106,7 +106,7 @@ private static DocumentField mutateDocumentField(DocumentField documentField) {
         if (isMetafield) {
             String metaField = randomValueOtherThanMany(excludeMetaFieldFilter,
-                () -> randomFrom(IndicesModule.getBuiltInMetadataFields()));
+                () -> randomFrom(MapperService.META_FIELDS_BEFORE_7DOT8));
```
Just as a last question: why is this necessary? I though the "old" collection of meta field names is only used when reading from "old" nodes? Why can't we use all built-in fields here since we're already on master?

mayya-sharipova replied:
@cbuescher I am struggling with BWC tests here, and I am not sure about the best approach.

`SearchHitsTests.testSerializationPre6_6_0` tests was failing if we use `IndicesModule.getBuiltInMetadataFields()`.

In this test we create a test SearchHit instance in 7.x and then do wire serialization to pre6_6 node and deserialization from it, and then assert that 7.x instance is equal to the deserialized instance from 6.x.  The test was failing if we use `IndicesModule.getBuiltInMetadataFields()` and include `_fied_names` as a meta-field in the created test instance.   As now during deserialization from pre 7.8 nodes we use `META_FIELDS_BEFORE_7DOT8` to decide if a field is meta-field, we put `_field_names` under document fields, which will not be equal to the original created test instance, where `_field_names` is under meta-fields.

Probably, the best approach would be to separate how we create test instances of SearchHit between pre 7.8 nodes and after 7.8 nodes.  What do you think?

cbuescher replied:
Are those particular tests needed on 7.x anymore? As far as I understand we'd never see a pre 6.6 node in a cluster running any 7x release at this point, 6.8 should be the last supported version? Maybe we can delete the test? I wonder if this uncovers other kinds of edge cases with the change though, so I'd prefer if we can sample "meta" fields from the "true" source (IndicesModule.getBuiltInMetadataFields()) and only fix tests where we need a specific pre-7.8 behaviour if really needed.
Mabye @javanna knows if `SearchHitsTests.testSerializationPre6_6_0` can be removed since he added it way back...

javanna replied:
https://github.com/elastic/elasticsearch/pull/36555 I guess the test (actually two tests) can be removed from 7.x given that 7.x nodes will never talk to pre 6.6 nodes.

mayya-sharipova replied:
@javanna Thanks, this would simplify things

mayya-sharipova replied:
@cbuescher Thanks for your suggestions. I have removed *pre6_6_0 tests, and also brought back `IndicesModule.getBuiltInMetadataFields()` for defining meta-fields in test items.

mayya-sharipova committed:
Remove pre 6.6 tests
